src/tests/unit_tests.c

```diff
@@ -0,0 +1,2116 @@
+// compiling for ASAN checking:
+//   gcc -o ../run/unit_tests -Wl,--no-undefined -O -g -fsanitize=address -fno-omit-frame-pointer -lasan -D_MISC_C_UNIT_TEST -D_JOHN_MISC_NO_LOG tests/unit_tests.c misc.c common.c
+
+// common type source to test functions:
+//	misc.c		(mostly done)
+//	common.c	(done)
+//	jumbo.c		(todo)
+//	list.c		(??)
+//	mask.c		(??)
+//	mask_ext.c	(??)
+//	memory.c    (??)
+//	unicode.c	(??)
+//	unicode_range.c (??)
+//	simd-intrinsics.c (??)
+//
+//  Likely could add tests for all hash types (or many).  Things like md2/4/5 sha/1/224/..512  sha3, etc, etc.
+//		These would be very fast set of known test vectors.  Make sure the functions return proper results, and
+//		simply list pass/fail.
+//
+//	base64_convert.c(may have it's own ST code already)
+//	compiler.c	(Possible unit test by building test suite of config files, then spawning john to test them)
+//	external.c	(Possible unit test by building test suite of config files, then spawning john to test them)
+//
+//	loader.c	(Not unit testable)
+//	logger.c	(Not unit testable)
+//	fuzz.c		(Not unit testable)
+//	inc.c		(Not unit testable)
+
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <time.h>
+#include <stdarg.h>
+#include <inttypes.h>
+#include "../misc.h"
+#include "../memory.h"
+#include "../common.h"
+
+#include "../sha2.h"
+
+
+// TOFIX after misc.[ch] and common.[ch] are fixed
+#ifdef _MSC_VER
+// code copied from jumbo.c for now, since it is not easy to 'use' jumbo.c by a
+// small external project.  I will 'patch' misc.c later and fix the problem,
+// but for now, this code is put into the unit_tests.c file.
+int vc_fixed_snprintf(char *Dest, size_t max_cnt, const char *Fmt, ...) {
+	va_list varg;
+	int len;
+
+	va_start(varg, Fmt);
+	len = _vsnprintf(Dest, max_cnt, Fmt, varg);
+	if (len < 0) {
+		int len_now = max_cnt;
+		Dest[max_cnt - 1] = 0;
+		// len = -1 right now.  We want to figure out exactly WHAT len should be, to stay C99 compliant.
+		while (len < 0) {
+			char *cp;
+			len_now *= 2;
+			cp = (char*)malloc(len_now);
+			len = _vsnprintf(cp, len_now, Fmt, varg);
+			free(cp);
+		}
+	}
+	va_end(varg);
+	return len;
+}
+#endif
+
+char *_fgetl_pad;
+#define _FGETL_PAD_SIZE 19000
+#define _ISHEX_CNT 260
+int _fgetl_fudge = 0; // if we miss a line, only show 1 error, we will have to add or skip a line (or more).
```

Comment:
I could bitch about the suggested line width of 80 when possible (**especially** for comments because they otherwise become damn near unreadable) but this is "your own" file anyway and I'm happy your back with us.